[READY] Making YCM run in Vim with only python3

Valloric opened the pull request:
Since https://github.com/Valloric/ycmd/pull/358, ycmd has supported Python 3. Now's the time to make the YCM client work in Python 3 too. As with the ycmd PR, **until we merge this, no other code-changing PR should be merged.** We need to get to a state where we have green Travis for Python 3 before any other PRs start landing.

This was substantially easier than porting ycmd, but it was still a massive pain. It's likely to fail in certain corner cases, so please give this a spin!

The best way to test it out is to build a Vim that has only Python 3 support. Our wiki page on [building Vim from source](https://github.com/Valloric/YouCompleteMe/wiki/Building-Vim-from-source) is a great start; you'll probably need to change `--enable-pythoninterp` to `--enable-python3interp` and similar for `--with-python-config-dir`. Or just get yourself a beta image of Ubuntu 16.04 LTS and run it in a VM (16.04 doesn't even ship Python 2 in the base image!). You can verify that you're running the right Vim by looking at the output of `vim --version`. It should have `-python` and `+python3`.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/valloric/youcompleteme/2016)
<!-- Reviewable:end -->

Valloric commented:
cc @vheon @micbou @puremourning

Valloric commented:
Review status: 0 of 30 files reviewed at latest revision, 2 unresolved discussions.

---

*[python/ycm/paths.py, line 38 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBejV-itTn5Ay_WfqY4:-KBejV-itTn5Ay_WfqY5:684837275) ([raw file](https://github.com/valloric/youcompleteme/blob/f8d95ab8c1d14cb36d8880f0ed4a6d03ef5ec986/python/ycm/paths.py#L38)):*
This is going to eventually break, but it's fine for about a decade. We should rewrite the checking logic at some point to not use a regex.

---

*[python/ycm/tests/vimsupport_test.py, line 1 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBekKxEjVP6AcM2fH9S:-KBekKxFWMk8wNLp1qfD:-1978430004) ([raw file](https://github.com/valloric/youcompleteme/blob/f8d95ab8c1d14cb36d8880f0ed4a6d03ef5ec986/python/ycm/tests/vimsupport_test.py#L1)):*
This is necessary because this file has non-ASCII chars.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016)*
<!-- Sent from Reviewable.io -->

Valloric commented:
@micbou I'm relying on you for a Windows support PR. :)

puremourning commented:
Awesome!

I just gave it a spin in py2 mode just to check for regressions, as it's a bit too late for building vim from source today. I just did a quick set of tests and most functionality that I can do quickly seems to work.

I noticed that sometimes exceptions are prefixed with `ycmd exception`. This doesn't seem to be universal. Is that the intention?

# Drive-by issue reports

## Vim traceback with `YcmToggleLogs`

Whenever I do `:YcmToggleLogs` I get the following traceback:

```
Error detected while processing function <SNR>132_ToggleLogs:
line    3:
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/ben/.vim/bundle/YouCompleteMe/autoload/../python/ycm/youcompleteme.py", line 591, in ToggleLogs
    self._CloseLogs( stdout = stdout, stderr = stderr )
  File "/Users/ben/.vim/bundle/YouCompleteMe/autoload/../python/ycm/youcompleteme.py", line 578, in _CloseLogs
    vimsupport.CloseBuffersForFilename( self._server_stdout )
  File "/Users/ben/.vim/bundle/YouCompleteMe/autoload/../python/ycm/vimsupport.py", line 869, in CloseBuffersForFilename
    "wiped out.".format( buffer_number, filename ) )
RuntimeError: Buffer -1 for filename '/var/folders/7k/qqmq40w11r3953ypw_jz2jjw0000gn/T/ycm_temp/server_58097_stdout.log' should already be wiped out.
```

This also affects `:YcmRestartServer` as it tries to clear the logs windows.

## Dodgy display of server addresses

Not sure if this is new or not, but I'm seeing this in `:YcmDebugInfo` for the Rust completer:

```
--   listening at: b'http://127.0.0.1:58104'
```

Maybe it is because ycmd specifically uses `bytes` for the address (as that's what is required by `requests`). Or something. So probably not related to this PR.

---

Review status: 0 of 34 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-:-KBelzUOeLZ1PkG1Apz9:-1584037693)*
<!-- Sent from Reviewable.io -->

Valloric commented:
@puremourning Thanks for giving it a go! 

Fix for the rust host string is https://github.com/Valloric/ycmd/pull/408. I'm looking into the log toggling issue.

---

Review status: 0 of 34 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-:-KBevGcrljt26xO_m5AB:-1075650709)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Fixed the log toggle issue. It's the same issue I've fixed in other places in this PR, and that's the use of the `is` operator for equality. `is` in Python checks for identity and _sometimes_ that works for equality when the other object is the same one, but that should not be relied upon. So let's try to nip this usage in reviews.

---

Review status: 0 of 34 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-:-KBf0qY8o6cYWeiYgiG3:-1508263576)*
<!-- Sent from Reviewable.io -->

Valloric commented:
We seem to be having some issues with [Python 2.6 on Mac on Travis](https://travis-ci.org/Valloric/YouCompleteMe/jobs/112505189). It's the only matrix row that fails. The absolute weirdest part is that it's the same Travis config that works for ycmd! 

@puremourning You probably have the best knowledge of OS X among us, could you take a look?

---

Review status: 0 of 35 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-:-KBfDNp05zIOkHK7OCyn:1814882788)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Ok, I remember now, we turned off Python 2.6 on Mac on Travis for ycmd because of the same problem. So I'm going to turn it off for ycmd as well. We still have 2.7 and 3.3, and no version of OS X we support uses 2.6 by default.

---

Review status: 0 of 35 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-:-KBfFdouaIEWyCh3PteA:43378790)*
<!-- Sent from Reviewable.io -->

micbou commented:
The `python-future` boilerplate is missing in `python/ycm/youcompleteme_test.py`.

---

Reviewed 28 of 30 files at r1, 2 of 2 files at r2, 3 of 4 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 1 of 2 files at r6, 1 of 1 files at r7.
Review status: 33 of 35 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[.travis.yml, line 56 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBevGNpTZuoP5ZRUP6w:-KBevGNpTZuoP5ZRUP6x:-919541355) ([raw file](https://github.com/valloric/youcompleteme/blob/96f8352fefd873cfe6296ea9686f205b24550434/.travis.yml#L56)):*
We don't need these caches except for pip and pyenv.

---

*[autoload/youcompleteme.vim, line 31 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBew_3m1UdoqvtOUDRR:-KBew_3n7x1Us9oFmgHX:-418272273) ([raw file](https://github.com/valloric/youcompleteme/blob/96f8352fefd873cfe6296ea9686f205b24550434/autoload/youcompleteme.vim#L31)):*
Missing blank line.

---

*[autoload/youcompleteme.vim, line 40 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBewa8wVy-Yo4Ba71my:-KBewa8wVy-Yo4Ba71mz:-418272273) ([raw file](https://github.com/valloric/youcompleteme/blob/96f8352fefd873cfe6296ea9686f205b24550434/autoload/youcompleteme.vim#L40)):*
Missing blank line.

---

*[autoload/youcompleteme.vim, line 44 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBewdsAOUWKZE9J_b00:-KBewdsBjqHuAusd1sHe:-418272273) ([raw file](https://github.com/valloric/youcompleteme/blob/96f8352fefd873cfe6296ea9686f205b24550434/autoload/youcompleteme.vim#L44)):*
Missing blank line.

---

*[python/ycm/paths.py, line 38 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBejV-itTn5Ay_WfqY4:-KBexF5LL-Qj_ncrIu6N:1311354079) ([raw file](https://github.com/valloric/youcompleteme/blob/f8d95ab8c1d14cb36d8880f0ed4a6d03ef5ec986/python/ycm/paths.py#L38)):*
We can use a range `[3-9]`.

---

*[python/ycm/paths.py, line 81 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBexY0tqG8M7RLoMGMw:-KBexY0uKWXET0Swmr0A:1587422256) ([raw file](https://github.com/valloric/youcompleteme/blob/96f8352fefd873cfe6296ea9686f205b24550434/python/ycm/paths.py#L81)):*
Maybe we should only search for `python` now (and remove the comment about Arch Linux).

---

*[python/ycm/tests/vimsupport_test.py, line 1182 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBezGKXZZk3fnAcijdf:-KBezGKYZZeKzsfe5E95:-490100116) ([raw file](https://github.com/valloric/youcompleteme/blob/96f8352fefd873cfe6296ea9686f205b24550434/python/ycm/tests/vimsupport_test.py#L1182)):*
Nice.

---

*[travis/travis_install.osx.sh, line 15 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBezWBL2URpDA5w2AHw:-KBezWBL2URpDA5w2AHx:-401414311) ([raw file](https://github.com/valloric/youcompleteme/blob/96f8352fefd873cfe6296ea9686f205b24550434/travis/travis_install.osx.sh#L15)):*
`node` and `go` can be removed.

---

*[travis/travis_install.sh, line 50 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBezhOKz1kv9_E6ANwq:-KBezhOLMeHDGDsbbpND:583631449) ([raw file](https://github.com/valloric/youcompleteme/blob/96f8352fefd873cfe6296ea9686f205b24550434/travis/travis_install.sh#L50)):*
Extra blank line.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-:-KBfFgFYssiIpazX8aS9:-1302316373)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Added missing boilerplate. BTW Reviewable has a "Start a new discussion" button which you can use to start a thread that needs to be resolved that's not connected to a specific file/line combo.

---

Review status: 30 of 36 files reviewed at latest revision, 10 unresolved discussions.

---

*[.travis.yml, line 56 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBevGNpTZuoP5ZRUP6w:-KBiFfLOT0gZoDighbZ3:-499146009) ([raw file](https://github.com/valloric/youcompleteme/blob/96f8352fefd873cfe6296ea9686f205b24550434/.travis.yml#L56)):*
Done.

---

*[autoload/youcompleteme.vim, line 31 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBew_3m1UdoqvtOUDRR:-KBiFmL9BwhECxs6lRJ9:-499146009) ([raw file](https://github.com/valloric/youcompleteme/blob/96f8352fefd873cfe6296ea9686f205b24550434/autoload/youcompleteme.vim#L31)):*
Done.

---

*[autoload/youcompleteme.vim, line 40 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBewa8wVy-Yo4Ba71my:-KBiFn3hyy7Bmr9wWm4D:-499146009) ([raw file](https://github.com/valloric/youcompleteme/blob/96f8352fefd873cfe6296ea9686f205b24550434/autoload/youcompleteme.vim#L40)):*
Done.

---

*[autoload/youcompleteme.vim, line 44 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBewdsAOUWKZE9J_b00:-KBiFpC9JrR1ejulA7Mp:-499146009) ([raw file](https://github.com/valloric/youcompleteme/blob/96f8352fefd873cfe6296ea9686f205b24550434/autoload/youcompleteme.vim#L44)):*
Done.

---

*[python/ycm/paths.py, line 38 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBejV-itTn5Ay_WfqY4:-KBiFw9g4V3pZb0pDhYY:-332542363) ([raw file](https://github.com/valloric/youcompleteme/blob/f8d95ab8c1d14cb36d8880f0ed4a6d03ef5ec986/python/ycm/paths.py#L38)):*
Good one! Done.

---

*[python/ycm/paths.py, line 81 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBexY0tqG8M7RLoMGMw:-KBiFzNFt0UGKkCWy74X:1202346254) ([raw file](https://github.com/valloric/youcompleteme/blob/96f8352fefd873cfe6296ea9686f205b24550434/python/ycm/paths.py#L81)):*
I was thinking about this, and we probably always want to pick Python 2 if available to ensure compatibility with old ycm_extra_conf files. I doubt many people wrote theirs to be compatible with py3 and they wouldn't like it if their setup "randomly" broke.

So I'm going to keep it as-is, but I've added more text to the comment explaining where we stand now.

---

*[travis/travis_install.osx.sh, line 15 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBezWBL2URpDA5w2AHw:-KBiGy5RKS4ruMdnc4LE:-499146009) ([raw file](https://github.com/valloric/youcompleteme/blob/96f8352fefd873cfe6296ea9686f205b24550434/travis/travis_install.osx.sh#L15)):*
Done.

---

*[travis/travis_install.sh, line 50 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBezhOKz1kv9_E6ANwq:-KBiH0IrSfXqqkbPsQNx:-499146009) ([raw file](https://github.com/valloric/youcompleteme/blob/96f8352fefd873cfe6296ea9686f205b24550434/travis/travis_install.sh#L50)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-:-KBiFKnUNvFviCiwX7AG:1238084554)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Well, out of the sort of morbid curiosity you only build up over a lifetime of looking at compile errors, I took a look at the the Python 2.6 error. And I can confidently say that I have literally no clue what it is about, except to say that it looks like python fails to build under pyenv under the Python 2.6 environment on OS X (which makes very little sense to me) and that it is either an OS X bug, or a pyenv bug.

Anyway, out of the (related) sort of morbid curiosity that comes from a lifetime of solving problems after finding them, I feel I am now duty-bound to at least fire up a OS X VM and see if i can repro :)

---

Review status: 26 of 38 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-:-KBo-scHBr1ahp4yfT-D:1545038588)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Of course the one version of OS X I don't have a pre-made snapshot for is Mavericks. Which I think is what Travis runs.

But fortunately I did create a snapshot containing all my "travis simulator" scripts, which is handy and I can reproduce trivially, so I'll see if I can work it out.

---

Review status: 26 of 38 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-:-KBo1hb75UCRDeJEv-5k:127522849)*
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 23 of 30 files at r1, 2 of 2 files at r2, 1 of 1 files at r7, 2 of 8 files at r9, 13 of 13 files at r10.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[python/ycm/omni_completer.py, line 118 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoCQCBRN818ofQyhYI:-KBoCQCCitSP3z23LPr3:-1014976950) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/python/ycm/omni_completer.py#L118)):*
Are we sure this error will not contain newlines? I see that we use it in other places so I think is not a problem :confused:

---

*[python/ycm/client/command_request.py, line 60 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoDFii0Y-mX2WudHSb:-KBoDFii0Y-mX2WudHSc:1422984132) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/python/ycm/client/command_request.py#L60)):*
same as above

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Well, I can confirm it only happens with `--enable-framework`. However, `--enable-shared` exhibits a different, even more cryptic error:

```
make: *** [libinstall] Error 1
```

Yep, that's it.

Sadly those are the only options that allow Python to work with dynamically loaded modules, such as those used by ycmd.

I seem recalled that we had similar issues with build on some 3.something versions so i tried python 2.6.9 (the latest) and that builds fine (under pyenv). However, under that version we get a link error when linking `ycm_core.so`. 

```
Undefined symbols for architecture x86_64:
  "_PyUnicodeUCS4_AsWideChar", referenced from:
      boost::python::converter::(anonymous namespace)::slot_rvalue_from_python<std::__1::basic_string<wchar_t, std::__1::char_traits<wchar_t>, std::__1::allocator<wchar_t> >, boost::python::converter::(anonymous namespace)::wstring_rvalue_from_python>::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in libBoostParts.a(builtin_converters.cpp.o)
  "_PyUnicodeUCS4_FromEncodedObject", referenced from:
      _encode_string_unaryfunc in libBoostParts.a(builtin_converters.cpp.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

Every version of 2.6 before that fails to even build python under pyenv afaict.

So, the solution might be to use python 2.6.9 on the test run (at least on OS X), and possibly also make some changes to our code or build (as yet undefined). @valloric I'll leave it up to you if you think we want to bump the minimum support (or at least _tested_) version of Python 2.6 just to test a (probably) obscure configuration of OS X. I'm not in love with testing different python 2.6 versions on different platforms, and I'm not totally convinced that testing on 2.6 on OS X is all that valuable in the long run.

I also don't have much of a clue how to fix the build error, which doesn't help :)

## In other news

I successfully built a version of vim with python3 and YCM seemed to work almost perfectly as far as I could tell with a quick spin. The only niggle was that i was using a `pyenv` and our "python detector" found the shim `python2` binary that `pyenv` has lying around. This lead to spam on my terminal and YCM claiming that it couldn't find python. I don't think that needs fixing really because it also helpfully told me to set `g:ycm_path_to_python_interpreter` which I did and it worked fine.

## Regression in YcmCompelter error

I remember this from a long time ago, but we've regressed the behaviour of exceptions with multiple lines. e.g. when you do something like `:YcmCompleter ThisIsNotValidCOmmand` you get something like:

```
ycmd exception: ValueError: Supported commands are:^@GoTo^@GoToDeclaration^@GoToDefinition^@StartServer^@StopServer^@See the docs for information on wha
t they do.
```

## Traceback when doing a `RefactorRename` and selecting `Quit` when prompted due to swap file

I haven't looked into this, and i'm about 100% certain I wrote the code below this, but I get a traceback executing the following steps under python 3 (sorry, I haven't tried under python 2).

- In Terminal1:

```
$ cd $YCMD/ycmd/tests/javascript/testdata
$ vim *.js
```

- In terminal 2:

```
$ cd $YCMD/ycmd/tests/javascript/testdata
$ vim file3.js
```

- put cursor on global variable and `:YcmCompleter RefactorRename testing`
- when prompted about swap file, hit `Q`

Observe:
```
Swap file "~/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/tests/javascript/testdata/.file1.js.swp" already exists!

Error detected while processing function <SNR>125_CompleterCommand:
line   18:
Traceback (most recent call last):
  File "/Users/ben/.vim/bundle/YouCompleteMe/autoload/../python/ycm/client/command_request.py", line 109, in _HandleFixitResponse
    vimsupport.ReplaceChunks( chunks )
  File "/Users/ben/.vim/bundle/YouCompleteMe/autoload/../python/ycm/vimsupport.py", line 635, in ReplaceChunks
    ( buffer_num, close_window ) = _OpenFileInSplitIfNeeded( filepath )
  File "/Users/ben/.vim/bundle/YouCompleteMe/autoload/../python/ycm/vimsupport.py", line 594, in _OpenFileInSplitIfNeeded
    'applied changes.'.format( filepath ) )
RuntimeError: Unable to open file: /Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/tests/javascript/testdata/file1.js
FixIt/Refactor operation aborted prior to completion. Your files have not been fully updated. Please use undo commands to revert the applied changes.
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/ben/.vim/bundle/YouCompleteMe/autoload/../python/ycm/youcompleteme.py", line 208, in SendCommandRequest
    return SendCommandRequest( arguments, completer )
  File "/Users/ben/.vim/bundle/YouCompleteMe/autoload/../python/ycm/client/command_request.py", line 130, in SendCommandRequest
    request.RunPostCommandActionsIfNeeded()
  File "/Users/ben/.vim/bundle/YouCompleteMe/autoload/../python/ycm/client/command_request.py", line 78, in RunPostCommandActionsIfNeeded
    return self._HandleFixitResponse()
  File "/Users/ben/.vim/bundle/YouCompleteMe/autoload/../python/ycm/client/command_request.py", line 111, in _HandleFixitResponse
    vimsupport.PostMultiLineNotice( e.message )
AttributeError: 'RuntimeError' object has no attribute 'message'
```

## Retries

I take it there isn't an upstream version of this which supports python3? TBH i didn't even realise this _wasn't_ a submodule!

## Todo

I wasn't able to easily build a Vim with *both* python2 and python3 bindings (again, because i'm using pyenv). I should be able to do this at some point, as I'm curious if there are any niggles.


That's about it. Sorry for the long ramble ... :)

---

Reviewed 23 of 30 files at r1, 2 of 2 files at r2, 1 of 1 files at r7, 2 of 8 files at r9, 13 of 13 files at r10.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[README.md, line 327 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoISJplaV1ABfnxE9p:-KBoISJplaV1ABfnxE9q:61067077) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/README.md#L327)):*
looks like there is an extra space between `i.e.` and `32`

---

*[README.md, line 486 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoIa-qeRKS_jbVKIpD:-KBoIa-qeRKS_jbVKIpE:-221396799) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/README.md#L486)):*
does this also return 1 if Vim is compiled with python/python3 in dynamic mode? I assume it does, just wondered if anyone happened to know.

---

*[autoload/youcompleteme.vim, line 35 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoIxVobsnGUQV0A4CH:-KBoIxVobsnGUQV0A4CI:-1477354207) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/autoload/youcompleteme.vim#L35)):*
Heh. You're about 100% right about that, though I can't think of a reason why anyone would care and/or notice. It should be transparent now that we totally decoupled the client and server (hoorah), right?

---

*[ci/appveyor/appveyor_install.bat, line 19 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoKEWQ7fw5Lr9Swv9P:-KBoKEWQ7fw5Lr9Swv9Q:-1326873517) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/ci/appveyor/appveyor_install.bat#L19)):*
bits -> bit

---

*[python/ycm/vimsupport.py, line 71 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoMvfdA1J-7uaLwQmH:-KBoMvfdA1J-7uaLwQmI:1423283772) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/python/ycm/vimsupport.py#L71)):*
bluerg. Does this mean that the type returned by `vim` module always require casting? That's way less than fun.

From a brief look at `:help python` Vim just claims to return a type of `String` which doesn't really help explain anything.

I think in your commit messages you were totally clear that internally, we should just play with `unicode` (aka py3 `str`) objects and convert at boundaries. It's just something (else) we're going to have to look out for in reviews I guess.

---

*[python/ycm/vimsupport.py, line 317 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoOeK8lejWFfaOh53v:-KBoOeK8lejWFfaOh53w:-169001645) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/python/ycm/vimsupport.py#L317)):*
anyone know what type is returned from `vim.eval`?

I'm just thinking that maybe we should be returning `ToUnicode( vim.eval( 'g:' ) )` here. Or maybe not - `vim.eval` probably returns some coerced type, and we have to handle it on a case-by-case basis.

Or does `ToUnicode` return the value unchanged if it isn't a string-like object?

---

*[python/ycm/vimsupport.py, line 322 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoPBodRokwJioP38K9:-KBoPBodRokwJioP38KA:-1198637292) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/python/ycm/vimsupport.py#L322)):*
Once again, forgive my poor understanding of these complexities, but on python 2, does `isinstance( u'unicode string', str )` return `True` ? In any case, should we return `ToUnicode( result )` from this if the returned value might be a py2 `str` or py3 `bytes` ? Again it comes down to what is returned by `vim.eval` which appears to be undocumented:

> 	- a string if the Vim expression evaluates to a string or number

---

*[python/ycm/vimsupport.py, line 912 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBogSnw-6UajHB1jm2M:-KBogSnw-6UajHB1jm2N:318337634) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/python/ycm/vimsupport.py#L912)):*
is this `str()` call safe in python3 ? Again, I don't really know and I'm too lazy to check... sorry.

---

*[python/ycm/client/command_request.py, line 60 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoDFii0Y-mX2WudHSb:-KBoiydIyrj7W4GuAUtd:-1761652447) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/python/ycm/client/command_request.py#L60)):*
Yeah, I think this causes the message to get a bit garbled.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-:-KBo6z3eK9WwB2GT9kqG:1312846242)*
<!-- Sent from Reviewable.io -->

Valloric commented:
@puremourning 

- python 2.6 on Mac: feel free to waste your time as much as you want :D. But it really is a waste. We don't support any OS X version which has 2.6 as the default. (I strive to support latest OS X and the one before, if that's not too much of a hassle.)
- I fixed the `:YcmCompleter` multi-line formatting.
- I _think_ I fixed the AttributError stuff (couldn't repro). Please confirm.
- retries: yeah it's literally one single file with about ~20 lines of code. It's from a GitHub gist, so there really isn't an "upstream" to talk to. It's not our copyright so it goes into `third_party`. Making it py3 compatible was a two-line change.
- I wouldn't recommend trying to build a Vim that supports both python2 and python3. It's not a very useful binary; you can't have both py2 and py3 initialized in the same process. I think with Vim's dynamic python loading you can get it to load _one_ of the two, but you'll always have to pick only one. As soon as you load both, BOOM. At least that was the case before. I doubt it has changed.
- May I recommend Reviewable's "Start a new discussion" button? :)

---

Review status: 32 of 38 files reviewed at latest revision, 12 unresolved discussions.

---

*[README.md, line 327 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoISJplaV1ABfnxE9p:-KBpFtxvwoiVILZDorTq:-499146009) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/README.md#L327)):*
Done.

---

*[README.md, line 486 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoIa-qeRKS_jbVKIpD:-KBpFyALibQDuyXNb5C4:-47376996) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/README.md#L486)):*
Maybe? Don't know. If it doesn't, then it didn't work before either. So I'm guessing it does since we use it in `plugin.vim`.

---

*[autoload/youcompleteme.vim, line 35 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoIxVobsnGUQV0A4CH:-KBpGAhsIxVFN7pK_gDz:-768214972) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/autoload/youcompleteme.vim#L35)):*
> I can't think of a reason why anyone would care and/or notice.

That's _exactly_ my point! Why would they care?! 

But someone will ask, and maybe, just _maybe_ they'll have a good argument why we should do it. 

---

*[ci/appveyor/appveyor_install.bat, line 19 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoKEWQ7fw5Lr9Swv9P:-KBpGvUtSMrOFCunFT94:-499146009) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/ci/appveyor/appveyor_install.bat#L19)):*
Done.

---

*[python/ycm/omni_completer.py, line 118 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoCQCBRN818ofQyhYI:-KBpH-QgtKV9b_Ux3dmT:1957997244) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/python/ycm/omni_completer.py#L118)):*
Fixed in `HandleServerException`.

---

*[python/ycm/vimsupport.py, line 71 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoMvfdA1J-7uaLwQmH:-KBpH6oRLuPhBMP3juux:921017915) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/python/ycm/vimsupport.py#L71)):*
Actually, `:help python` is correct, but doesn't explain in detail. 

I spent some time looking into this while porting. All vim Python APIs that return a string-like object always return `str` objects. So on py2, that's the equivalent of a py3 `bytes` object. On py3, it's the equivalent of a py2 `unicode`.

Rather interestingly, all vim Python APIs that _accept_ a string-like object transparently accept `str`, `bytes` and `unicode` on whatever Python version. I shudder to think which encoding they assume for the non-unicode versions of those objects; probably the buffer encoding from the vim `encoding` setting.

---

*[python/ycm/vimsupport.py, line 317 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoOeK8lejWFfaOh53v:-KBpI7NaUnpW_JEdYUv6:1886694754) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/python/ycm/vimsupport.py#L317)):*
`vim.eval` can return dicts, ints, strings etc. So can't call `ToUnicode` here because that will call `str()` on objects that aren't string-like.

---

*[python/ycm/vimsupport.py, line 322 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoPBodRokwJioP38K9:-KBpJ8nvI3IeBSWUJjEM:59389460) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/python/ycm/vimsupport.py#L322)):*
You are indeed correct kind sir, this needed fixing! Done.

---

*[python/ycm/vimsupport.py, line 912 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBogSnw-6UajHB1jm2M:-KBpN3WDk6ruVmlog_mA:291566230) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/python/ycm/vimsupport.py#L912)):*
It should be. Note that our `str` is patched by Python.

---

*[python/ycm/client/command_request.py, line 60 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoDFii0Y-mX2WudHSb:-KBpNDgPI8C4TUJfcM6A:-499146009) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/python/ycm/client/command_request.py#L60)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-:-KBpBVpsK4J33MzR_vlg:-1737255164)*
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 6 of 6 files at r11.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016)*
<!-- Sent from Reviewable.io -->

micbou commented:
I tested the four client/server combinations (`py2/py2`, `py2/py3`, `py3/py2`, and `py3/py3`) and they all worked. I just needed to set the `g:ycm_path_to_python_interpreter` when the Python versions were different (`py2/py3` and `py3/py2`), which is expected behavior.

So this is <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> for me with an update to latest ycmd (maybe merge PR Valloric/ycmd#410 before).

---

Reviewed 5 of 8 files at r9, 13 of 13 files at r10, 6 of 6 files at r11.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[README.md, line 486 \[r10\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBoIa-qeRKS_jbVKIpD:-KBt_jUZEnKaX3yKy8yr:1272606920) ([raw file](https://github.com/valloric/youcompleteme/blob/2ebde118c7b4c160c92a9f1908d293b4afcd0092/README.md#L486)):*
I can confirm it works (on Windows).

---

*[python/ycm/tests/vimsupport_test.py, line 1 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBekKxEjVP6AcM2fH9S:-KBtg75VQw-iASMp1XBb:-2040087109) ([raw file](https://github.com/valloric/youcompleteme/blob/f8d95ab8c1d14cb36d8880f0ed4a6d03ef5ec986/python/ycm/tests/vimsupport_test.py#L1)):*
I would prefer without the `-*-`s. They are for Emacs.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-:-KBtibnWnigTGKAPIB7P:-1651755492)*
<!-- Sent from Reviewable.io -->

Valloric commented:
@micbou Thanks for the thorough testing! I updated this to latest ycmd branch after merging https://github.com/Valloric/ycmd/pull/410.

---

Review status: 36 of 38 files reviewed at latest revision, 9 unresolved discussions.

---

*[python/ycm/tests/vimsupport_test.py, line 1 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KBekKxEjVP6AcM2fH9S:-KBuEfDk5srTaaqRxA-S:-499146009) ([raw file](https://github.com/valloric/youcompleteme/blob/f8d95ab8c1d14cb36d8880f0ed4a6d03ef5ec986/python/ycm/tests/vimsupport_test.py#L1)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-:-KBuEMjR62p1qpaWO3Hj:-1054506890)*
<!-- Sent from Reviewable.io -->

esphen commented:
I'm trying to compile this branch to test the featureset with my python3-only vim, currently getting `Traceback (most recent call last):` in my editor window when opening vim.

My install steps are as follows. My vimrc can be found [here](https://github.com/esphen/vimrc).
```bash
# Open vim and do :PluginInstall! for good luck
cd .vim/bundle/YouCompleteMe
git checkout py3
git submodule update --init --recursive
./install.py --tern-completer
```

Running `:YcmToggleLogs` gives me the following
```
Traceback (most recent call last):                                                                                                                                                                                   
  File "/usr/lib64/python3.4/runpy.py", line 170, in _run_module_as_main                                                                                                                                             
    "__main__", mod_spec)                                                                                                                                                                                            
  File "/usr/lib64/python3.4/runpy.py", line 85, in _run_code                                                                                                                                                        
    exec(code, run_globals)                                                                                                                                                                                          
  File "/home/espen/.vim/bundle/YouCompleteMe/python/ycm/../../third_party/ycmd/ycmd/__main__.py", line 181, in <module>                                                                                             
    Main()                                                                                                                                                                                                           
  File "/home/espen/.vim/bundle/YouCompleteMe/python/ycm/../../third_party/ycmd/ycmd/__main__.py", line 158, in Main                                                                                                 
    if not CompatibleWithCurrentCoreVersion():                                                                                                                                                                       
  File "/home/espen/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/server_utils.py", line 50, in CompatibleWithCurrentCoreVersion                                                                                   
    import ycm_core                                                                                                                                                                                                  
ImportError: dynamic module does not define init function (PyInit_ycm_core)
```

Any steps you would like me to do for debugging? Was I not able to switch branch correctly?

Valloric commented:
@esphen The error you got means that the YCM native libs have been compiled for Python 3 but you are loading them into a Python 2 process (or the other way around).

esphen commented:
@Valloric yeah, that works, thanks :) Didn't even have to delete the YCM checkout. Installed `python3-devel` and compiled using `python3 install.py --tern-completer`. Worked like a charm. If I see any bugs, I'll report them to you guys

Valloric commented:
@esphen I added an entry in the FAQ for the `ImportError` issue you encountered.

@puremourning @vheon If one of you could give an LGTM this weekend that would be awesome. :)

puremourning commented:
> python 2.6 on Mac: feel free to waste your time as much as you want :D. But it really is a waste. We don't support any OS X version which has 2.6 as the default. (I strive to support latest OS X and the one before, if that's not too much of a hassle.)

Yeah. The moment passed. :grimacing: 

> I fixed the :YcmCompleter multi-line formatting.

Confirm resolved :)

> I think I fixed the AttributError stuff (couldn't repro). Please confirm.

Confirmed in both py2 and py3

> I wouldn't recommend trying to build a Vim that supports both python2 and python3. It's not a very useful binary; you can't have both py2 and py3 initialized in the same process. I think with Vim's dynamic python loading you can get it to load one of the two, but you'll always have to pick only one. As soon as you load both, BOOM. At least that was the case before. I doubt it has changed.

With reference to [this issue](https://github.com/Valloric/YouCompleteMe/issues/1818) and [a command I made](https://github.com/Valloric/YouCompleteMe/pull/1821#issuecomment-163384057) on its subsequent fix.

Can't remember if I mentioned this, but I was a bit confused about this. I remembered an issue about UltiSnips running in py3 where I suggested it was because YCM was running under py2 and UltiSnips under py3. I wanted to try it out,and finally got round to it today. However I'm not able to be sure if it is working or not, as I'm seeing a traceback with diagnostics (I've raised a new issue for this)

>  May I recommend Reviewable's "Start a new discussion" button? :)

Certainly...

---

Reviewed 3 of 6 files at r11, 1 of 1 files at r12, 1 of 1 files at r13, 1 of 1 files at r15.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KC6JZxjp7XwkBAY1LMr:-KC6JZxjp7XwkBAY1LMs:-9306902) (no related file):*
I'm getting a traceback when YCM places signs.

```
"cpp/ycm/Candidate.cpp" 120L, 3465C
Error detected while processing function <SNR>108_OnInsertLeave[6]..<SNR>108_OnFileReadyToParse:
line    9:
Traceback (most recent call last):
Error detected while processing function <SNR>108_OnInsertLeave[6]..<SNR>108_OnFileReadyToParse:
line    9:
  File "<string>", line 1, in <module>
Error detected while processing function <SNR>108_OnInsertLeave[6]..<SNR>108_OnFileReadyToParse:
line    9:
  File "/Users/ben/.vim/bundle/YouCompleteMe/autoload/../python/ycm/youcompleteme.py", line 507, in HandleFileParseRequest
Error detected while processing function <SNR>108_OnInsertLeave[6]..<SNR>108_OnFileReadyToParse:
line    9:
    self.UpdateDiagnosticInterface()
Error detected while processing function <SNR>108_OnInsertLeave[6]..<SNR>108_OnFileReadyToParse:
line    9:
  File "/Users/ben/.vim/bundle/YouCompleteMe/autoload/../python/ycm/youcompleteme.py", line 490, in UpdateDiagnosticInterface
Error detected while processing function <SNR>108_OnInsertLeave[6]..<SNR>108_OnFileReadyToParse:
line    9:
    self._diag_interface.UpdateWithNewDiagnostics( self._latest_diagnostics )
Error detected while processing function <SNR>108_OnInsertLeave[6]..<SNR>108_OnFileReadyToParse:
line    9:
  File "/Users/ben/.vim/bundle/YouCompleteMe/autoload/../python/ycm/diagnostic_interface.py", line 79, in UpdateWithNewDiagnostics
Error detected while processing function <SNR>108_OnInsertLeave[6]..<SNR>108_OnFileReadyToParse:
line    9:
    _UpdateSquiggles( self._buffer_number_to_line_to_diags )
Error detected while processing function <SNR>108_OnInsertLeave[6]..<SNR>108_OnFileReadyToParse:
line    9:
  File "/Users/ben/.vim/bundle/YouCompleteMe/autoload/../python/ycm/diagnostic_interface.py", line 113, in _UpdateSquiggles
Error detected while processing function <SNR>108_OnInsertLeave[6]..<SNR>108_OnFileReadyToParse:
line    9:
    vimsupport.ClearYcmSyntaxMatches()
Error detected while processing function <SNR>108_OnInsertLeave[6]..<SNR>108_OnFileReadyToParse:
line    9:
  File "/Users/ben/.vim/bundle/YouCompleteMe/autoload/../python/ycm/vimsupport.py", line 205, in ClearYcmSyntaxMatches
Error detected while processing function <SNR>108_OnInsertLeave[6]..<SNR>108_OnFileReadyToParse:
line    9:
    if match[ 'group' ].startswith( 'Ycm' ):
Error detected while processing function <SNR>108_OnInsertLeave[6]..<SNR>108_OnFileReadyToParse:
line    9:
  File "/Users/ben/.pyenv/versions/2.7.6/Python.framework/Versions/2.7/lib/python2.7/site-packages/future/types/newstr.py", line 118, in __getitem__
Error detected while processing function <SNR>108_OnInsertLeave[6]..<SNR>108_OnFileReadyToParse:
line    9:
    return newstr(super(newstr, self).__getitem__(y))
Error detected while processing function <SNR>108_OnInsertLeave[6]..<SNR>108_OnFileReadyToParse:
line    9:
TypeError: string indices must be integers
```

I repro'd this on my main mac and a clean OS X VM under different Vim versions. The clean one from the VM (at vim head):

Vim version (this happens under python3, python2 and python2 with python3 dynamically available too):

```
os-x-yosemite:ycmd ben$ vim --version
VIM - Vi IMproved 7.4 (2013 Aug 10, compiled Mar  1 2016 23:10:15)
MacOS X (unix) version
Included patches: 1-1468
Compiled by ben@os-x-yosemite.shared
Huge version without GUI.  Features included (+) or not (-):
+acl             +farsi           +mouse_netterm   +tag_binary
+arabic          +file_in_path    +mouse_sgr       +tag_old_static
+autocmd         +find_in_path    -mouse_sysmouse  -tag_any_white
-balloon_eval    +float           +mouse_urxvt     -tcl
-browse          +folding         +mouse_xterm     +terminfo
++builtin_terms  -footer          +multi_byte      +termresponse
+byte_offset     +fork()          +multi_lang      +textobjects
+channel         -gettext         -mzscheme        +title
+cindent         -hangul_input    +netbeans_intg   -toolbar
-clientserver    +iconv           +packages        +user_commands
+clipboard       +insert_expand   +path_extra      +vertsplit
+cmdline_compl   +job             -perl            +virtualedit
+cmdline_hist    +jumplist        +persistent_undo +visual
+cmdline_info    +keymap          +postscript      +visualextra
+comments        +langmap         +printer         +viminfo
+conceal         +libcall         +profile         +vreplace
+cryptv          +linebreak       -python          +wildignore
-cscope          +lispindent      +python3/dyn     +wildmenu
+cursorbind      +listcmds        +quickfix        +windows
+cursorshape     +localmap        +reltime         +writebackup
+dialog_con      -lua             +rightleft       -X11
+diff            +menu            -ruby            -xfontset
+digraphs        +mksession       +scrollbind      -xim
-dnd             +modify_fname    +signs           -xsmp
-ebcdic          +mouse           +smartindent     -xterm_clipboard
+emacs_tags      -mouseshape      +startuptime     -xterm_save
+eval            +mouse_dec       +statusline      -xpm
+ex_extra        -mouse_gpm       -sun_workshop
+extra_search    -mouse_jsbterm   +syntax
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  fall-back for $VIM: "/Users/ben/roots/vim-py3/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H   -DMACOS_X_UNIX  -g -O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1
Linking: gcc   -L/usr/local/lib -o vim        -lm -lncurses  -liconv -framework Cocoa
os-x-yosemite:ycmd ben$
```

I can get this behaviour simply by opening `third_party/ycmd/ycmd/tests/clang/testdata/FixIt_Clang_cpp11.cpp`

my HEAD is `7b9b5ac6883148115100e3419dda3e035966b7f7` and i'm clean.

Just me? I'm certain this wasn't the case when I tested it before, and I don't recall seeing a change here which would cause it :/

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-:-KC5yXzN3qsSZw1dAdlQ:2054687261)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KC6JZxjp7XwkBAY1LMr:-KC6PJw-Mt94Q-j9pOS8:-779652749) (no related file):*
Actually I think the issue is with the change to `VimExpressionToPythonType`. It looks like `ClearYcmSyntaxMatches` expects it to return an iterable (`vim.eval` returns a list in this case).

It also looks like the check is wonky.

---

*[python/ycm/vimsupport.py, line 322 \[r15\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KC6QxZgu4AnR4_xNXfI:-KC6QxZgu4AnR4_xNXfJ:-1726075120) ([raw file](https://github.com/valloric/youcompleteme/blob/7b9b5ac6883148115100e3419dda3e035966b7f7/python/ycm/vimsupport.py#L322)):*
i think it needs to be something like:

```python
def VimExpressionToPythonType( vim_expression ):
  """Returns a Python type from the return value of the supplied Vim expression.
  If the expression returns a list, dict or other non-string type, then it is
  returned unmodified. If the string return can be converted to an
  integer, returns an integer, otherwise returns the result converted to a
  Unicode string."""

  result = vim.eval( vim_expression )
  if not ( isinstance( result, str ) or isinstance( result, bytes ) ):
    return result

  try:
    return int( result )
  except ValueError:
    return ToUnicode( result )
```

or any De Morgan equivalent of the if statement, depending on preference :grinning: 

This fixes the error for me in py2 and py3, though I don't use `ycm_extra_conf_vim_data` which I think is the other place this method is used.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016)*
<!-- Sent from Reviewable.io -->

puremourning commented:
With the fix for the issue below, this <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-:-KC6U7mG_5xu80nlk9cG:1041455227)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Reviewed 2 of 6 files at r11.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Review status: 36 of 38 files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KC6JZxjp7XwkBAY1LMr:-KC6stOiCOI5pmcns1rF:1233504890) (no related file):*
Can confirm & repro. Fixed with your recommended code for `VimExpressionToPythonType`.

---

*[python/ycm/vimsupport.py, line 322 \[r15\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-KC6QxZgu4AnR4_xNXfI:-KC6zwwQcA1ik8KTSPYF:1125850155) ([raw file](https://github.com/valloric/youcompleteme/blob/7b9b5ac6883148115100e3419dda3e035966b7f7/python/ycm/vimsupport.py#L322)):*
Took your code & added a bunch of tests. Thanks!

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016)*
<!-- Sent from Reviewable.io -->

Valloric commented:
@puremourning Thanks for taking another look! :)

---

Review status: 36 of 38 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-:-KC7-8JoUlw949JwcIi7:-814428078)*
<!-- Sent from Reviewable.io -->

puremourning commented:
No problem at all :)

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r16.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-:-KC7-Sr5_dunkRm6wJy9:322464044)*
<!-- Sent from Reviewable.io -->

Valloric commented:
I'm going to pull the trigger on this. It's Saturday morning here, so I should have plenty of time to address any issues that come up if people report them right away.

@homu r=micbou

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/2016#-:-KC7-HIIz3uPA3BehSsb:1768228690)*
<!-- Sent from Reviewable.io -->

homu commented:
:pushpin: Commit 0bcecb1 has been approved by `micbou`

<!-- @homu r=micbou 0bcecb14d89d6dc23c3d2a136cd607fc5040b5e8 -->

homu commented:
:zap: Test exempted - [status](https://travis-ci.org/Valloric/YouCompleteMe/builds/113926007)
